_posts/2017-10-04-Law-of-quality.md

```diff
@@ -0,0 +1,106 @@
+
+This post examines the difficulties in these three matters to shed some light into why it's actually quite difficult. The intention is not deal with these subjects a in thorough manner - that would fill a book I'm not qualified to write.
+
+## Dates
+
+Date and time are something familiar to us, but the devil is in the details. [The Long, Painful History of Time](http://naggum.no/lugm-time.html) gives a good account of the intricacies of date and time.
+
+Provided you understand this historical context and time zones, this is only half of the victory. Let's consider Java and a standard JDBC interface to a relation database, say PostgreSQL. 
+
+Originally Java had two Date classes: ```java.util.Date``` and ```java.sql.Date```. Despite Java core developers being smart, it
+turned out that poor ```java.util.Date``` had issues. A decade later Java time API was rewritten in [JSR 310](https://community.oracle.com/docs/DOC-983209). And then [JodaTime](http://www.joda.org/joda-time/) came to rescue the suffering Java community. Now the confused developer has deprecated (and broken) legacy classes laying around, some newer ones to choose from and a lot of headache. 
+
+Okay, but surely the situation is better in SQL databases? We'll define our table like this:
+```
+CREATE TABLE dadas
+  (
+    le_time TIMESTAMPT NOT NULL
+  ) ;
+```
+
+Oops, we just failed. According to [PostgreSQL documentation](https://www.postgresql.org/docs/9.1/static/datatype-datetime.html) this leaves out timezone, which means effectively system locale. 
+
+```
+The SQL standard requires that writing just timestamp be equivalent to timestamp without 
+time zone, and PostgreSQL honors that behavior. (Releases prior to 7.3 treated it as 
+timestamp with time zone.)
+```
+
+It is inadvisable to assume that system default locale in the database server is the same as in the application server, unless you somehow explicitly control it. But let's be gracious and assume you have everything under control in the backend. What about the user interface? Enter Javascript and let the mortals tremble.
+
+```
+new Date(2012,12,12)
+ > Sat Jan 12 2013 00:00:00 GMT+0200 (EET)
+new Date(2012,0,1)
+ > Sun Jan 01 2012 00:00:00 GMT+0200 (EET)
+new Date(2012,-1,1)
+ > Thu Dec 01 2011 00:00:00 GMT+0200 (EET)
+```
+
+Javascript is kind enough to allow data slip without warnings or exceptions like that. You get dates, just maybe a bit different dates from what you expected. And what might be the timezone should you do something like that? You can't rely on the user's browser to have any specific timezone set even if you can control the backend server. 
```

Comment:
“slip **through**”?

“maybe just a bit different than what you expected”

“what might the timezone be”